File: antia_numerical_methods@Beliavsky/c/fun/crouth.c
/*	Solution of a system of linear equations using Crout's algorithm 
            with iterative refinement

	N : (input) Number of equations to be solved
	NUM : (input) Number of different sets (each with N equations) of
	         equations to be solved
	A : (input) The matrix of coefficient of size LJ*N
	         A[i][j] is the coefficient of x_j in ith equation
	B : (output) Array of size LJ*N containing triangular decomposition of matrix A
	X : (input/output) The matrix containing right hand sides (size LJ*NUM)
	        X[j][i] is the ith element of jth right hand side
			at output it will contain the solutions
	DET, IDET : (output) The determinant of the matrix = DET*2**IDET
	INC : (output) Integer array of length N containing information about
			interchanges performed during elimination
	LJ : (input) Second dimension of arrays A and X in calling function
	REPS : (input) Required relative precision in solution
	IFLG : (input) Integer parameter which determines the type of computation
		required.
		If IFLG<=0, both elimination and solution are calculated
			and IFLG is set to 2
		If IFLG=1, only elimination is done and IFLG is set to 2
		If IFLG>=2 only solution is calculated, the triangular
		    decomposition should have been calculated earlier
	WK : (output) Array of length NUM containing the estimated errors.
		WK[I] will contain estimated error in the Ith solution.
		
	Error status is returned by the value of the function CROUTH.
		0 value implies successful execution
		11 implies that iterative refinement did not converge
		102 implies (N<=0 or N>LJ) 
		122 implies some pivot turned out to be zero and hence
			matrix must be nearly singular

	Required functions : CROUT
*/

#include <math.h>
#include <stdlib.h>

int crout(int n, int num, double *a, double *x, double *det, int *idet,
	int inc[], int lj, int *iflg);

int crouth(int n, int num, double *a, double *b, double *x, double *det,
	int *idet, int inc[], int lj, double reps, int *iflg, double wk[])

{
	int i,j,k,it,iflg1,ier,num1, nitr=10;
	double r1,r2,rp1,rp;
/* If long double is supported, use that for residuals */
	long double d1,d2;
	double *wk1, *wk2;

	if(n<=0 || n>lj) return 102;

	if(*iflg<2) {
/*	Preserving the matrix for calculating the residuals */
		for(i=0; i<n; ++i) {
			for(j=0; j<n; ++j) b[j*lj+i]= a[j*lj+i];
		}
		iflg1=1;
/*	Perform LU decomposition using crout */
		ier=crout(n,num,b,x,det,idet,inc,lj,&iflg1);
		if(ier>100) return ier;
	}

	if(*iflg==1) {*iflg=2; return 0;}
	*iflg=2;
	num1=1;
	wk1=(double *) calloc((size_t) n, sizeof(double));
	wk2=(double *) calloc((size_t) n, sizeof(double));

/*	Solving the systems with NUM different right-hand sides */
	for(j=0; j<num; ++j) {
		for(i=0; i<n; ++i) {
			wk1[i]= x[j*lj+i];
/*	Preserving the RHS for calculating residuals */
			wk2[i]= x[j*lj+i];
			x[j*lj+i]=0.0;
		}

		rp1=0.0;
/*	The iterative refinement */
		for(it=1; it<= nitr ; ++it) {
			ier=crout(n,num1,b,wk1,det,idet,inc,lj,&iflg1);
			if(ier>100) {free(wk2); free(wk1); return ier;}
			r1=0.0; r2=0.0;
			for(i=0; i<n; ++i) {
				if(fabs(wk1[i])>r1) r1=fabs(wk1[i]);
				x[j*lj+i]=x[j*lj+i]+wk1[i];
				if(fabs(x[j*lj+i])>r2) r2=fabs(x[j*lj+i]);
			}

/*	The error estimate  */
			if(it==2) wk[j]=r1;
			if(r2==0.0) goto nextit;
			rp=r1/r2;
			if(rp<reps) goto nextit;
			if(rp>rp1 && it>1) {ier=11; goto nextit;}
			rp1=rp;

/*	Calculating the residue  */
			for(i=0; i<n; ++i) {
				d1= wk2[i];
				for(k=0; k<n; ++k) {
					d2=a[i*lj+k];
					d1=d1-d2*x[j*lj+k];
				}
				wk1[i]=d1;
			}
		}
		ier=11;
nextit: ;
	}
	free(wk2); free(wk1);
	return ier;
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub8/crouth.f
!	Solution of a system of linear equations using Crout's algorithm 
!            with iterative refinement
!
!	N : (input) Number of equations to be solved
!	NUM : (input) Number of different sets (each with N equations) of
!	         equations to be solved
!	A : (input) The matrix of coefficient of size LJ*N
!	         A(I,J) is the coefficient of x_J in Ith equation
!	B : (output) Array of size LJ*N containing triangular decomposition of matrix A
!	X : (input/output) The matrix containing right hand sides (size LJ*NUM)
!	        X(I,J) is the Ith element of Jth right hand side
!	     	at output it will contain the solutions
!	DET, IDET : (output) The determinant of the matrix = DET*2**IDET
!	INC : (output) Integer array of length N containing information about
!		interchanges performed during elimination
!	LJ : (input) First dimension of arrays A and X in calling program
!	REPS : (input) Required relative precision in solution
!	IER : (output) Error flag, IER=0 signifies successful execution
!		IER=11 implies that iterative refinement did not converge
!		IER=102 implies (N.LE.0 or N.GT.LJ) 
!		IER=122 implies some pivot turned out to be zero and hence
!			matrix must be nearly singular
!	IFLG : (input) Integer parameter which determines the type of computation
!		required.
!		If IFLG.LE.0, both elimination and solution are calculated
!			and IFLG is set to 2
!		If IFLG=1, only elimination is done and IFLG is set to 2
!		If IFLG.GE.2 only solution is calculated, the triangular
!		    	decomposition should have been calculated earlier
!	WK : Scratch array of length (2*N+NUM)
!		at output WK(I) will contain estimated error in Ith solution
!
!	Required routines : CROUT

      SUBROUTINE CROUTH(N,NUM,A,B,X,DET,IDET,INC,LJ,REPS,IER,IFLG,WK)
      IMPLICIT REAL*8(A-H,O-Z)
      PARAMETER(NITR=10)
!	If the compiler supports higher precision use this statement
!	Otherwise the iterative refinement process may not converge
!	as it is necessary to use higher precision arithmetic while
!	calculating the residuals.
      REAL*16 D1,D2
      DIMENSION A(LJ,N),B(LJ,N),X(LJ,NUM),INC(N),WK(2*N+NUM)

      IF(N.LE.0.OR.N.GT.LJ) THEN
        IER=102
        RETURN
      ENDIF

      IF(IFLG.LE.1) THEN
!	Preserving the matrix for calculating the residuals
        DO 1000 I=1,N
          DO 1000 J=1,N
1000    B(J,I)=A(J,I)
        IFLG1=1
!	Perform LU decomposition using CROUT
        CALL CROUT(N,NUM,B,X,DET,IDET,INC,LJ,IER,IFLG1,WK)
        IF(IER.GT.100) RETURN
      ENDIF

      IF(IFLG.EQ.1) THEN
        IFLG=2
        RETURN
      ENDIF
      IFLG=2
      NUM1=1
      IER=0

!	Solving the systems with NUM different right-hand sides
      DO 5000 J=1,NUM
        DO 2000 I=1,N
          WK(I)=X(I,J)
!	Preserving the RHS for calculating residuals
          WK(I+N)=WK(I)
          X(I,J)=0.0
2000    CONTINUE

        RP1=0
!	The iterative refinement
        DO 3000 IT=1,NITR
          CALL CROUT(N,NUM1,B,WK,DET,IDET,INC,LJ,IER1,IFLG1,WK)
          R1=0.0
          R2=0.0
          DO 2200 I=1,N
            IF(ABS(WK(I)).GT.R1) R1=ABS(WK(I))
            X(I,J)=X(I,J)+WK(I)
            IF(ABS(X(I,J)).GT.R2) R2=ABS(X(I,J))
2200      CONTINUE

!	The error estimate
          IF(IT.EQ.2) WK(J+2*N)=R1
          IF(R2.EQ.0.0) GO TO 5000
          RP=R1/R2
          IF(RP.LT.REPS) GO TO 5000
          IF(RP.GT.RP1.AND.IT.GT.1) THEN
            IER=11
            GO TO 5000
          ENDIF
          RP1=RP

!	Calculating the residue
          DO 2600 I=1,N
            D1=WK(I+N)
            DO 2400 K=1,N
!              D1=D1-A(I,K)*X(K,J)
!	To force double length accumulation of residuals
             D2=A(I,K)
             D1=D1-D2*X(K,J)
2400        CONTINUE
            WK(I)=D1
2600      CONTINUE
3000    CONTINUE
        IER=11

5000  CONTINUE
!	The error estimates
      DO 6000 I=1,NUM
6000  WK(I)=WK(I+2*N)
      END
